Fix recent Safer CPP bot regressions

csaavedra opened the pull request:
#### 278f3a4e1280352da1b79cdad794c8409aa2ea34
<pre>
Fix recent Safer CPP bot regressions
<a href="https://bugs.webkit.org/show_bug.cgi?id=294238">https://bugs.webkit.org/show_bug.cgi?id=294238</a>

Reviewed by Chris Dumez.

* Source/JavaScriptCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:
* Source/JavaScriptCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/JavaScriptCore/parser/ModuleAnalyzer.h:
* Source/WebCore/Modules/mediastream/CanvasCaptureMediaStreamTrack.cpp:
(WebCore::CanvasCaptureMediaStreamTrack::Source::captureCanvas):
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::effectiveViewTransitionName):
* Source/WebCore/html/FormAssociatedElement.h:
(WebCore::FormAssociatedElement::asProtectedHTMLElement const):
* Source/WebCore/html/FormController.cpp:
(WebCore::FormController::ownerForm):
* Source/WebCore/html/FormListedElement.cpp:
(WebCore::FormListedElement::resetFormAttributeTargetObserver):
(WebCore::FormListedElement::name const):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::shouldUpdateAccessibilityRegions const):
* Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
(-[WebScrollerImpDelegateMac mouseLocationInScrollerForScrollerImp:]):
(-[WebScrollerImpDelegateMac scrollerImp:animateKnobAlphaTo:duration:]):
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.h:
(WebCore::ScrollerPairMac::checkedVerticalScroller):
(WebCore::ScrollerPairMac::checkedVerticalScroller const):
(WebCore::ScrollerPairMac::checkedHorizontalScroller):
(WebCore::ScrollerPairMac::checkedHorizontalScroller const):
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
(-[WebScrollerImpPairDelegateMac scrollerImpPair:convertContentPoint:toScrollerImp:]):
(WebCore::ScrollerPairMac::init):
(WebCore::ScrollerPairMac::~ScrollerPairMac):
(WebCore::ScrollerPairMac::updateValues):
(WebCore::ScrollerPairMac::releaseReferencesToScrollerImpsOnTheMainThread):
(WebCore::ScrollerPairMac::scrollbarStateForOrientation const):
(WebCore::ScrollerPairMac::mouseIsInScrollbar):
(WebCore::ScrollerPairMac::setUseDarkAppearance):
(WebCore::ScrollerPairMac::setScrollbarWidth):
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateFromStateNode):
* Source/WebCore/rendering/svg/RenderSVGResourceContainer.cpp:
(WebCore::RenderSVGResourceContainer::idChanged):
* Source/WebCore/style/values/color/StyleLightDarkColor.cpp:
(WebCore::Style::toStyleColor):
* Source/WebKit/UIProcess/API/Cocoa/WKNavigationResponse.mm:
(-[WKNavigationResponse dealloc]):
(-[WKNavigationResponse _navigationInitiatingFrame]):
* Source/WebKit/UIProcess/API/Cocoa/_WKThumbnailView.mm:
(-[_WKThumbnailView _viewWasParented]):
* Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm:
(-[WKTextAnimationManager textPreviewsForChunk:completion:]):
* Source/WebKit/WebProcess/WebCoreSupport/WebCryptoClient.cpp:
* Source/WebKitLegacy/mac/WebCoreSupport/TextIndicatorWindow.mm:
(TextIndicatorWindow::setTextIndicator):

Canonical link: <a href="https://commits.webkit.org/296189@main">https://commits.webkit.org/296189@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9160b2137996dc5975b5aa0b38d8b86569eea205

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/107640 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27318 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17731 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/112853 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58181 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/109599 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28010 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/35831 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81731 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110572 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22206 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97021 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62111 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21640 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15160 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/57616 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/100227 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91578 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15194 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/115966 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/106184 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34714 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25593 "Found 60 new test failures: accessibility/mac/element-focus.html compositing/repaint/sticky-repaint-on-scroll.html compositing/scrolling/async-overflow-scrolling/mac/overflow-in-flex-empty-tiles.html css3/scroll-snap/resnap-after-layout.html css3/scroll-snap/scroll-snap-discrete-wheel-event-in-mainframe.html fast/events/pointer-events-2.html fast/events/scroll-in-scaled-page-with-overflow-hidden.html fast/events/wheel/wheel-event-destroys-frame.html fast/events/wheel/wheel-event-in-passive-region-non-cancelable.html fast/events/wheel/wheel-event-listeners-on-document-made-passive.html ... (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90769 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35091 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93271 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90515 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23080 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35433 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13208 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/30470 "Build is being retried. Recent messages:") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/34624 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40180 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/130499 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34370 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35476 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37731 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36033 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebKit/UIProcess/API/Cocoa/WKNavigationResponse.mm:
```diff
@@ -39,7 +39,8 @@ - (void)dealloc
 
-    _navigationResponse->~NavigationResponse();
+    // FIXME: This RefPtr should not be necessary. Remove it once clang static analyzer is fixed.
+    RefPtr { _navigationResponse.get() }->~NavigationResponse();
```
There should be a specific bug on file that is referenced from the FIXMEs, for each such unique analyzer issue, ideally assigned to @rniwa.

csaavedra replied:
TBH I copied this from a different place in the same file, so I assumed this is a known issue. @rniwa ?

rniwa replied:
Why is RefPtr not necessary here?

csaavedra replied:
Actually now that I look at this more in detail I made wrong assumptions. I'll remove the comments. Also please note that this is still marked as a draft :-)

annevk commented on Source/WebCore/html/FormController.cpp:
```diff
@@ -43,7 +43,7 @@ HTMLFormElement* FormController::ownerForm(const FormListedElement& control)
     // state during parsing and form owners of such controls might be
     // indeterminate.
-    return control.asHTMLElement().hasAttributeWithoutSynchronization(HTMLNames::formAttr) ? nullptr : control.form();
+    return CheckedRef { control.asHTMLElement() }->hasAttributeWithoutSynchronization(HTMLNames::formAttr) ? nullptr : control.form();
```
It seems wrong to use CheckedRef where Ref can be used. Probably `asHTMLElement` needs to have a `protectedAsHTMLElement` variant that returns a Ref.

cdumez replied:
Right, we may as well as use a Ref instead of a Ref here since it's possible. There is no difference in performance but Ref guarantees lifetime while CheckedRef just crashes if we get lifetime wrong. May as well not take the risk. If we want to introduce a new getter like @annevk suggests, I'd go with `asProtectedHTMLElement()` instead. It sounds a bit better and would be consistent with the existing `ensureProtectedFoo()` functions we have.

csaavedra replied:
Thanks for the suggestions, makes sense!

annevk commented on Source/WebCore/page/scrolling/mac/ScrollerPairMac.h:
```diff
@@ -94,6 +94,9 @@ class ScrollerPairMac : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<S
     ScrollerMac& horizontalScroller() { return m_horizontalScroller; }
 
+    CheckedRef<ScrollerMac> checkedVerticalScroller() { return m_verticalScroller; }
+    CheckedRef<ScrollerMac> checkedHorizontalScroller() { return m_horizontalScroller; }
```
```suggestion
    ScrollerMac& verticalScroller() { return m_verticalScroller; }
    CheckedRef<ScrollerMac> checkedVerticalScroller() { return m_verticalScroller; }
    ScrollerMac& horizontalScroller() { return m_horizontalScroller; }
    CheckedRef<ScrollerMac> checkedHorizontalScroller() { return m_horizontalScroller; }
```
This makes it clearer the comment still applies.

cdumez commented on Source/JavaScriptCore/bytecode/StructureStubClearingWatchpoint.h:
```diff
@@ -76,7 +76,7 @@ class StructureTransitionStructureStubClearingWatchpoint final : public Watchpoi
 
 private:
-    PolymorphicAccessJITStubRoutine* m_owner;
+    RefPtr<PolymorphicAccessJITStubRoutine> m_owner;
```
cc @Constellation and @kmiller68 as this one is a bit concerning. Usually concerning a raw pointer into a RefPtr and is a recipe for introducing reference cycles and thus leaks. I didn't look into the details here. I'll let a JSC engineer review this part.

Constellation replied:
This is wrong, which causes massive memory leaking.

csaavedra replied:
OK, I would use a CheckedPtr here but I don't see these widely used in JSC so I will revert this change and update the expectations for JSC, I leave this to JSC people to fix.

cdumez commented on Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
```diff
@@ -325,14 +325,14 @@ - (void)scrollerImpPair:(NSScrollerImpPair *)scrollerImpPair updateScrollerStyle
 String ScrollerPairMac::scrollbarStateForOrientation(ScrollbarOrientation orientation) const
 {
-    return orientation == ScrollbarOrientation::Vertical ? m_verticalScroller.scrollbarState() : m_horizontalScroller.scrollbarState();
+    return orientation == ScrollbarOrientation::Vertical ? CheckedRef { m_verticalScroller }->scrollbarState() : CheckedRef { m_horizontalScroller }->scrollbarState();
```
Any reason we're not using `checkedVerticalScroller()` here?

csaavedra replied:
Wrong constness.

cdumez replied:
We can make the getter non-const I think?

csaavedra replied:
You mean, to make `::scrollbarStateForOrientation()` non-const?

cdumez replied:
I thought you said you could not call `checkedVerticalScroller()` because of wrong constness? I think this code should call `checkedVerticalScroller()`, fixing the constness of `checkedVerticalScroller()` if necessary.

csaavedra replied:
Ah, now I understand, right, I'll add const variants (we need both).

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/_WKThumbnailView.mm:
```diff
@@ -220,7 +220,7 @@ - (void)_viewWasParented
 
     if (!_exclusivelyUsesSnapshot) {
-        auto obscuredContentInsets = _webPageProxy->obscuredContentInsets();
+        auto obscuredContentInsets = RetainPtr { _webPageProxy.get() }->obscuredContentInsets();
```
Using RetainPtr here seems wrong and will likely crash. We should be using RefPtr for a WebPageProxy.

cdumez reviewed (changes requested):
r- because of the RetainPtr issue. We also need to check with JSC about the extra RefPtr data member.

Constellation reviewed (changes requested):
r-

cdumez commented on Source/JavaScriptCore/bytecode/ObjectPropertyConditionSet.cpp:
```diff
@@ -177,7 +177,7 @@ void ObjectPropertyConditionSet::dumpInContext(PrintStream& out, DumpContext* co
     out.print("[");
     if (m_data)
-        out.print(listDumpInContext(*m_data, context));
+        out.print(listDumpInContext(*m_data.copyRef(), context));
```
I think we can mark `m_data` as `const` and use `lazyInitialize()` to lazily initialize it instead. This is JSC so let's try and avoid ref-counting as much as possible, in case this is perf sensitive code.

csaavedra replied:
It looks to me that `ObjectPropertyConditionSet` objects are reassigned in a few places, so I'm not sure that the `lazyInitialize()` trick will work here (IIUC, making `m_data` `const` deletes the assignment operator, so re-assigning is not possible). Don't want to enter that rabbit hole and delay this patch longer, so I will leave these changes out of the MR.

(that being said, it seems to me that this `dumpInContext()` method is used for logging of some kind, so I'm not sure it's really performance-sensitive, but I agree with you on it being JSC it's better to be cautious, so I'll just leave these changes out).

cdumez commented on Source/JavaScriptCore/parser/ModuleAnalyzer.h:
```diff
@@ -26,6 +26,7 @@
 #pragma once
 
 #include "ErrorType.h"
+#include "JSModuleRecord.h"
```
Is this really needed?

csaavedra replied:
I think the include is needed for the "ForwardDeclChecker" warning. https://build.webkit.org/results/Apple-Sequoia-Safer-CPP-Checks/296153%40main%20%283833%29/scan-build-output/StaticAnalyzerUnexpectedRegressions/JavaScriptCore/StaticAnalyzerReports/report-ModuleAnalyzer.h-unknown-0-63856d.html#EndPath

cdumez commented on Source/WebCore/page/scrolling/mac/ScrollerPairMac.h:
```diff
@@ -92,7 +92,11 @@ class ScrollerPairMac : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<S
 
     // Only for use by WebScrollerImpPairDelegateMac. Do not use elsewhere!
     ScrollerMac& verticalScroller() { return m_verticalScroller; }
+    CheckedRef<ScrollerMac> checkedVerticalScroller()  { return m_verticalScroller; }
```
nit: there is an extra space here.

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebCryptoClient.cpp:
```diff
@@ -29,6 +29,7 @@
 #include "WebPageProxyMessages.h"
 #include "WebProcess.h"
 #include "WebProcessProxyMessages.h"
+#include <WebCore/CryptoKeyData.h>
```
Is this really needed?

csaavedra replied:
Same about the forward declaration warning.

cdumez replied:
Oh I see, I forgot about those warnings.

cdumez reviewed (approved):
LGTM with fixes.

csaavedra committed:
Fix recent Safer CPP bot regressions
https://bugs.webkit.org/show_bug.cgi?id=294238

Reviewed by Chris Dumez.

* Source/JavaScriptCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:
* Source/JavaScriptCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/JavaScriptCore/parser/ModuleAnalyzer.h:
* Source/WebCore/Modules/mediastream/CanvasCaptureMediaStreamTrack.cpp:
(WebCore::CanvasCaptureMediaStreamTrack::Source::captureCanvas):
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::effectiveViewTransitionName):
* Source/WebCore/html/FormAssociatedElement.h:
(WebCore::FormAssociatedElement::asProtectedHTMLElement const):
* Source/WebCore/html/FormController.cpp:
(WebCore::FormController::ownerForm):
* Source/WebCore/html/FormListedElement.cpp:
(WebCore::FormListedElement::resetFormAttributeTargetObserver):
(WebCore::FormListedElement::name const):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::shouldUpdateAccessibilityRegions const):
* Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
(-[WebScrollerImpDelegateMac mouseLocationInScrollerForScrollerImp:]):
(-[WebScrollerImpDelegateMac scrollerImp:animateKnobAlphaTo:duration:]):
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.h:
(WebCore::ScrollerPairMac::checkedVerticalScroller):
(WebCore::ScrollerPairMac::checkedVerticalScroller const):
(WebCore::ScrollerPairMac::checkedHorizontalScroller):
(WebCore::ScrollerPairMac::checkedHorizontalScroller const):
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
(-[WebScrollerImpPairDelegateMac scrollerImpPair:convertContentPoint:toScrollerImp:]):
(WebCore::ScrollerPairMac::init):
(WebCore::ScrollerPairMac::~ScrollerPairMac):
(WebCore::ScrollerPairMac::updateValues):
(WebCore::ScrollerPairMac::releaseReferencesToScrollerImpsOnTheMainThread):
(WebCore::ScrollerPairMac::scrollbarStateForOrientation const):
(WebCore::ScrollerPairMac::mouseIsInScrollbar):
(WebCore::ScrollerPairMac::setUseDarkAppearance):
(WebCore::ScrollerPairMac::setScrollbarWidth):
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateFromStateNode):
* Source/WebCore/rendering/svg/RenderSVGResourceContainer.cpp:
(WebCore::RenderSVGResourceContainer::idChanged):
* Source/WebCore/style/values/color/StyleLightDarkColor.cpp:
(WebCore::Style::toStyleColor):
* Source/WebKit/UIProcess/API/Cocoa/WKNavigationResponse.mm:
(-[WKNavigationResponse dealloc]):
(-[WKNavigationResponse _navigationInitiatingFrame]):
* Source/WebKit/UIProcess/API/Cocoa/_WKThumbnailView.mm:
(-[_WKThumbnailView _viewWasParented]):
* Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm:
(-[WKTextAnimationManager textPreviewsForChunk:completion:]):
* Source/WebKit/WebProcess/WebCoreSupport/WebCryptoClient.cpp:
* Source/WebKitLegacy/mac/WebCoreSupport/TextIndicatorWindow.mm:
(TextIndicatorWindow::setTextIndicator):

Canonical link: https://commits.webkit.org/296189@main
